iroh-bytes/src/store/file.rs

```diff
@@ -0,0 +1,2483 @@
+pub(crate) struct PathOptions {
+    /// Path to the directory where data and outboard files are stored.
+    data_path: PathBuf,
+    /// Path to the directory where temp files are stored.
+    /// This *must* be on the same device as `data_path`, since we need to
+    /// atomically move temp files into place.
+    temp_path: PathBuf,
+}
+
+impl PathOptions {
+    fn new(root: &Path) -> Self {
+        Self {
+            data_path: root.join("data"),
+            temp_path: root.join("temp"),
+        }
+    }
+
+    fn owned_data_path(&self, hash: &Hash) -> PathBuf {
+        self.data_path.join(format!("{}.data", hash.to_hex()))
+    }
+
+    fn owned_outboard_path(&self, hash: &Hash) -> PathBuf {
+        self.data_path.join(format!("{}.obao4", hash.to_hex()))
+    }
+
+    fn owned_sizes_path(&self, hash: &Hash) -> PathBuf {
+        self.data_path.join(format!("{}.sizes4", hash.to_hex()))
+    }
+
+    fn temp_file_name(&self) -> PathBuf {
+        self.temp_path.join(temp_name())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct Options {
+    path: PathOptions,
+    /// Inline storage options.
+    inline: InlineOptions,
+}
+
+#[derive(derive_more::Debug)]
+pub(crate) enum ImportSource {
+    TempFile(PathBuf),
+    External(PathBuf),
+    Memory(#[debug(skip)] Bytes),
+}
+
+impl ImportSource {
+    fn content(&self) -> MemOrFile<&[u8], &Path> {
+        match self {
+            Self::TempFile(path) => MemOrFile::File(path.as_path()),
+            Self::External(path) => MemOrFile::File(path.as_path()),
+            Self::Memory(data) => MemOrFile::Mem(data.as_ref()),
+        }
+    }
+
+    fn len(&self) -> io::Result<u64> {
+        match self {
+            Self::TempFile(path) => std::fs::metadata(path).map(|m| m.len()),
+            Self::External(path) => std::fs::metadata(path).map(|m| m.len()),
+            Self::Memory(data) => Ok(data.len() as u64),
+        }
+    }
+}
+
+/// An entry for a partial or complete blob in the store.
+#[derive(Debug, Clone, derive_more::From)]
+pub struct Entry(BaoFileHandle);
+
+impl super::MapEntry for Entry {
+    fn hash(&self) -> Hash {
+        self.0.hash()
+    }
+
+    fn size(&self) -> BaoBlobSize {
+        let size = self.0.current_size().unwrap();
+        tracing::info!("redb::Entry::size() = {}", size);
+        BaoBlobSize::new(size, self.is_complete())
+    }
+
+    fn is_complete(&self) -> bool {
+        self.0.is_complete()
+    }
+
+    async fn available_ranges(&self) -> io::Result<bao_tree::ChunkRanges> {
+        todo!()
+    }
+
+    async fn outboard(&self) -> io::Result<impl Outboard> {
+        self.0.outboard()
+    }
+
+    async fn data_reader(&self) -> io::Result<impl AsyncSliceReader> {
+        Ok(self.0.data_reader())
+    }
+}
+
+impl super::MapEntryMut for Entry {
+    async fn batch_writer(&self) -> io::Result<impl BaoBatchWriter> {
+        Ok(self.0.writer())
+    }
+}
+
+#[derive(derive_more::Debug)]
+pub(crate) struct Import {
```

Comment:
I have a lot of places where I use content for a HashAndFormat because hash_and_format is too verbose and hash is weird because then you need to do `hash.hash()`.

I guess source is a reasonable name for ImportSource.

At some point I called the HashAndFormat thing `cid`, but then people would hate me because that is overloaded in ipfs land...

Reply:
Ok, how about `content_id`?